### DISCUSSION TITLE
Having difficulty using the Link component with the asChild property and custom component as the child element

### DISCUSSION THREAD

**Discussion by AlexMachin1997** [expo/router#632] [Q&A]
Hi everyone,

When using the Link component with the "asChild" to forward the props to my custom component/s, however, I'm very confused about how you should actually use this component.

When I wrap my components in the Link component and set asChild to true I get the following warning: "Function components cannot be given refs. Attempts to access this ref will fail warning"

**What I did to try and resolve the issue:**
1. Add the forwardRef to the component the Link component is wrapped around, I add the ref to the element e.g. I'm using TouchableRipple from react-native-paper so that was attached to that. However, when I tried out the change the original warning still persisted
2. Removed the ref, everything works and no warning appears. which is very confusing.

It's not very clear why this is happening as based on the documentation there is no mention of needing to add the forwardRef in the Link section of the documentation;

It would be great to get some clarifications around the usage of the Link component as I favour this for navigating instead of using onClick onLongPress etc

Thank's,

Alex

Warning screenshot:
![Screenshot_20230604-180748](https://github.com/expo/router/assets/29917195/e32c2b6d-0b92-48a9-8994-c4a7d069f6c9)

**Comment by dmrcnr** [comment]
If you could share your code, the problem would be more understandable, but did you try innerRef instead of forwardRef?

**Comment by shyankashani** [comment]
Any updates on this? I'm having the exact same issue.

**Reply by AlexMachin1997** [reply]
Sadly no I've not had time to look into it further yet I've been super busy with work stuff and wool be until the start of next month.

Will try and have another look when I get chance 🤞

**Comment by FlamurMMT** [comment]
I think react-native-paper has a PR to address this, not sure if it's been merged in yet. 

Personally I have a custom button component and I include the ref and wrap the component with forwardref so for example:

```
import { forwardRef } from 'react'
import { TouchableHighlight } from 'react-native'

function Button({ children, type = 'primary', disabled, text, ...rest }, ref) {
      <TouchableHighlight
           ref={ref}
           ....
           ....
      />
      ....
}

export default forwardRef(Button)
```

**Comment by toppsdown** [comment]
Getting set up with expo and RN for the first time. As a test, I extracted the button to a new component and ran into this same issue. I get an error in the console saying: _"Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?"_ and the link stops working.

### Working code:

```jsx
import { Link } from "expo-router"
import { View, Button, StyleSheet } from "react-native"

export default function TestScreen() {
  return (
    <View style={styles.container}>
      <Link href="/home" asChild>
        <Button title="Go Home" color="#841584" />
      </Link>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#6544E9",
    alignItems: "center",
    justifyContent: "center",
  },
  image: {
    width: 165,
    height: 50,
  },
})
```

### broken code:
```jsx
import { Link } from "expo-router"
import { View, Button, StyleSheet } from "react-native"

const MyButton = ({ title, color }) => {
  return <Button title={title} color={color} />
}

export default function TestScreen() {
  return (
    <View style={styles.container}>
      <Link href="/home" asChild>
        <MyButton title="Go Home" color="#841584" />
      </Link>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#6544E9",
    alignItems: "center",
    justifyContent: "center",
  },
  image: {
    width: 165,
    height: 50,
  },
})
```

**Reply by marklawlor** [reply]
You need to follow the instructions in React's error message

https://react.dev/reference/react/forwardRef

```tsx
const MyButton = fowardRef(({ title, color }) => {
  return <Button title={title} color={color} />
})
```

**Comment by LucastheFront** [comment]
Hi everyone,

I'm also having the same problem trying to use my own Button component inside a Link component from expo router.
Using forwardRef, I don't have any warnings but the button is not routing me to the desired path. Nothing is happening.

Here is my code:

### Button Component

`
function Button(props, ref) {
return (
    <Pressable ref={ref}>
      <Text>{props.children}</Text>
    </Pressable>
  );
}  
export default forwardRef(Button);
`
### App Component:
`<Link href="/new-recipe" asChild><Button primary>Créer une recette</Button></Link>`

Any ideas on how to solve this ? When I use the useRef hook inside the App Component and console log it, I can see that the ref is successfully set up on the Pressable.

Thanks for your help,
Lucas

**Reply by elenitaex5** [reply]
I fix this wrapping my custom component into a `<Pressable />` Like this:
```
<Link href="/new-recipe" asChild>
    <Pressable>
        <Button primary>Créer une recette</Button>  //or other custom component
    </Pressable>
</Link>
```

**Comment by AlexMachin1997** [comment]
Hi everyone,

I've just had another look at this after upgrading to expo-router version 2, however, I'm still having issues when wanting to wrap a custom component in a Link for navigating.

It would be great if the documentation actually covered how this can be achieved. I don't get the warning but the navigation isn't triggered so it's not been passed correctly or it's missing certain props to enable this functionality.

**Comment by almustaphasilvester** [comment]
Hi everyone, 

You need to pass the onPress and onClick props to your custom button as such. 

```
import { Pressable, Image} from 'react-native';
import { GlobalStyles } from '../assets/styles/GlobalStyling.js';

export default function NextButton({ isDisabled, onPress, onClick }) {

    const disabled = require('../assets/iOS/1X/next_disabled.png');
    const enabled = require('../assets/iOS/1X/next_button.png');


    return (
        <Pressable onPress={onPress} onClick={onClick}>
            <Image source={isDisabled ? disabled : enabled} style={GlobalStyles.nextButton}/>
        </Pressable>
    )
};
```

The Link prop requires those two props to be passed down.

**Reply by almustaphasilvester** [reply]
Had to revise this method as it wasn't working consistently and ended up creating the component with Link wrapped around the pressable and passing the "href" as a prop.

**Reply by th0r0nd0r** [reply]
For anyone still coming across this discussion, I was able to get things working with a combination of this solution and using `forwardRef`. I also just went ahead and passed all parent props to my component with `...rest`, rather than just `onPress` and `onClick`.

```javascript
import { forwardRef } from 'react';
import { Pressable } from 'react-native';
import MyComponent from './myComponent';

export const MyButton = forwardRef(({ myCustomProp, ..rest }, ref) => {
    return (
        <Pressable {...rest} ref={ref}>
            <MyComponent myCustomProp />
        </Pressable>
    );
});
```

```javascript
import { Link } from 'expo-router';
import { MyButton } from './myButton';

export default function MyScreen() {
    return (
        <Link href="my-path" asChild>
            <MyButton myCustomProp />
        </Link>
    )
}
```

And it's working so far in the iOS sim and on web.

**Comment by flanger001** [comment]
Here is how I handled this:

```jsx
import { Pressable, Text, View } from "react-native";
import { Link } from "expo-router";
import styles, { colors } from "../utils/styles";
import { forwardRef } from "react";
import BaseView from "../components/BaseView";

const LinkButton = forwardRef(({ onPress, onClick, ...props }, ref) => (
  <Pressable
    style={({ pressed }) => (
      {
        ...styles.button.base,
        ...(pressed ? styles.button.active : styles.button.passive)
      }
    )}
    ref={ref}
    onPress={onPress}
    onClick={onClick}
  >
    {({ pressed }) => (
      <Text
        style={{
          ...styles.fonts.base,
          textAlign: "center",
          color: (pressed ? colors.white.primary : colors.blue.primary)
        }}
      >
        {props.children}
      </Text>
    )}
  </Pressable>)
)

const Root = () => {
  return (
    <View style={styles.baseView}>
      <Text style={styles.textLanding}>Hello</Text>
      <Link href="/login" asChild>
        <LinkButton>
          Login
        </LinkButton>
      </Link>
    </View>
  )
}

export default Root;
```

Looks great:
![IMG_2918](https://github.com/expo/router/assets/1645017/b93996d5-e987-4d7b-8f2b-28201171b3d7)

Everything works:
![IMG_2919](https://github.com/expo/router/assets/1645017/47342770-660a-418c-beda-fb575db29168)

**Comment by Danny-Pye** [comment]
Set asChild={true} tells the Link component to treat its child component as the clickable element. Just put a Pressable inside the Link, that works for me.


<Link asChild={true} href={{ pathname: "/", params: {id: 1} }}>
    <Pressable>
        <YourComponent />
    </Pressable>
</Link>

**Comment by hugobayoud** [comment]
This answer from another topic seems to work well for me with a good use of typescript  : https://github.com/expo/router/issues/278#issuecomment-1656025761

**Comment by elenitaex5** [comment]
If anyone needs to fix this... 
I had my custom component and even wrapping with link this, it only worked in web, not android nor iOS. Wrapping my custom component inside a `<Pressable />` too fix it.

```
<Link href={linkHref} asChild>
    <Pressable>
        <MyCustomComponent />
    </Pressable>
</Link>
```

Hope it helps

**Comment by EraYaN** [comment]
What also a bit of a shame is that

```tsx
<Link href={`/details/${id}`} asChild>
  <Text style={styles.linkText}>{`Show details for ${${id}}`}</Text>
</Link>
```
does work (no warning, working link)
```tsx
<Link href={`/details/${id}`} asChild>
  <Button title={`Show details for ${${id}}`}/>
</Link>
```
does not. (warning, sometimes working link but breaks component render)

**Comment by noahjhaub** [comment]
Any news on this?
I'm also having this issue. For me the button is just not showing. It's there but just completely transparent.
And I can't figure out how to fix this in a good way.

I have a custom Button component which uses `Pressable` and forwards the ref:
```
type Props = { pressableStyle?: string; textStyle?: string; children: React.ReactNode } & PressableProps;

export const Button = forwardRef<View, Props>(({ pressableStyle, textStyle, children, ...pressableProps }, ref) => {
    return (
        <Pressable style={tw.style("bg-white py-4 justify-center items-center rounded-xl w-full", pressableStyle)} {...pressableProps} ref={ref}>
            {typeof children === "string" ? <Text style={tw.style("text-black font-semibold text-base", textStyle)}>{children}</Text> : children}
        </Pressable>
    );
});
```

And I use it with `<Link  />` like this:
```
<Link href={"/onboarding"} asChild>
    <Button pressableStyle="bg-black" textStyle="text-white" onPress={() => console.log(`🟠🟠🟠 touch`)}>
        Los geht's
    </Button>
</Link>
```

Only way I could make it work was my wrapping the custom button component in a `Pressable` where I use it. But this just feels very weird. I basically then have a `Pressable` wrapped in a `Pressable` and I hate that.

**Comment by mbellamj** [comment]
Hello,

I also face this kind of issue. I fixed it like this

```
interface ButtonProps extends Pick<PressableProps, 'onPress' | 'onLongPress'> {
  children?: ReactNode;
  style?: StyleProp<ViewStyle>;
}

export const Button = forwardRef<View, ButtonProps>(
  ({ children, style, ...rest }, ref) => {
    const colorScheme = useColorScheme() ?? 'light';
    const pressedColor = Colors[colorScheme].buttonPressed;

    return (
      <Pressable
        ref={ref}
        style={({ pressed }) => [
          { backgroundColor: pressed ? pressedColor : 'transparent' },
          styles.button,
          style,
        ]}
        {...rest}>
        {children}
      </Pressable>
    );
  },
);

Button.displayName = 'Button';
```

**Comment by torin-st** [comment]
My workaround:

`import { router } from 'expo-router';`
`import { Button } from 'react-native';`
`...`
`<Button title='Add' onPress={ () => router.navigate("/add") }/>`
`...`

**Reply by JayBee007** [reply]
Yes, this works well, but I noticed that the default back button after replacing on a Stack is not present. The forwardRef solution worked for me.
